lib/slave/src/ESMStates.cc

```diff
@@ -0,0 +1,238 @@
+                }
+            }
+        }
+
+        // Invalid state request
+        if ((control.get_requested_state() == State::SAFE_OP) or (control.get_requested_state() == State::OPERATIONAL))
+        {
+            return Context::build(State::INIT, StatusCode::INVALID_REQUESTED_STATE_CHANGE);
+        }
+
+        // BOOTSTRAP not supported yet. If implemented, need to check the SII to know if enabled.
+        if (control.get_requested_state() == State::BOOT)
+        {
+            return Context::build(id_, StatusCode::BOOTSTRAP_NOT_SUPPORTED);
+        }
+
+        return Context::build(State::INIT);
+    }
+
+    PreOP::PreOP(AbstractESC& esc, PDO& pdo)
+        : AbstractState(State::PRE_OP, esc, pdo)
+    {
+    }
+
+    void PreOP::on_entry(Context, Context)
+    {
+        if (mbx_)
+        {
+            mbx_->set_sm_activate(true);
+        }
+        pdo_.set_sm_output_activated(false);
+        pdo_.set_sm_input_activated(false);
+    }
+
+    Context PreOP::routine_internal(Context, ALControl control)
+    {
+        if (mbx_ and not mbx_->is_sm_config_ok())
+        {
+            return Context::build(State::INIT, StatusCode::INVALID_MAILBOX_CONFIGURATION_PREOP);
+        }
+
+        if (control.get_requested_state() == State::SAFE_OP)
+        {
+            if (pdo_.configure_pdo_sm() != hresult::OK)
+            {
+                return Context::build(id_);
+            }
+
+            StatusCode pdo_sm_config_status_code = pdo_.is_sm_config_ok();
+            if (pdo_sm_config_status_code != StatusCode::NO_ERROR)
+            {
+                return Context::build(id_, pdo_sm_config_status_code);
+            }
+
+            return Context::build(State::SAFE_OP);
+        }
+
+        if (control.get_requested_state() == State::INIT)
+        {
+            return Context::build(INIT);
+        }
+
+        if (control.get_requested_state() == State::OPERATIONAL or control.get_requested_state() == State::BOOT)
+        {
+            return Context::build(id_, StatusCode::INVALID_REQUESTED_STATE_CHANGE);
+        }
+
+        return Context::build(id_);
+    }
+
+
+    SafeOP::SafeOP(AbstractESC& esc, PDO& pdo)
+        : AbstractState(State::SAFE_OP, esc, pdo)
+    {
+    }
+
+    void SafeOP::on_entry(Context oldStatus, Context newStatus)
+    {
+        if (oldStatus.get_state() == State::OPERATIONAL and newStatus.al_status_code != StatusCode::NO_ERROR)
+        {
+            pdo_.set_sm_output_activated(false);
+        }
+        else
+        {
+            pdo_.set_sm_output_activated(true);
+            pdo_.set_sm_input_activated(true);
+        }
+    }
+
+    Context SafeOP::routine_internal(Context currentStatus, ALControl control)
+    {
+        pdo_.update_process_data_input();
+        pdo_.update_process_data_output();
+
+        if (mbx_ and not mbx_->is_sm_config_ok())
+        {
+            return Context::build(State::INIT, StatusCode::INVALID_MAILBOX_CONFIGURATION_PREOP);
```

Comment:
INVALID_MAILBOX_CONFIGURATION_PREOP ? from safeop ?

Reply:
Yep... in the standard. verified with the ctt